src/main/java/com/google/devtools/build/lib/rules/java/JavaConfiguration.java

```diff
@@ -218,6 +223,58 @@ private static void checkLegacyToolchainFlagIsUnset(String flag, Label label)
     }
   }
 
+  private static void checkLanguageAndRuntimeVersion(@Nullable String rawLanguageVersion,
+      @Nullable String rawRuntimeVersion, boolean isExec) throws InvalidConfigurationException {
+    int languageVersion;
+    if (Strings.isNullOrEmpty(rawLanguageVersion)) {
+      return;
+    }
+    try {
+      languageVersion = Integer.parseUnsignedInt(rawLanguageVersion);
+    } catch (NumberFormatException e) {
+      // This is unexpected, but for the sake of compatibility with unusual toolchains, we should
+      // not fail the build.
+      return;
+    }
+
+    int runtimeVersion;
+    if (Strings.isNullOrEmpty(rawRuntimeVersion)) {
+      return;
+    }
+    int lastUnderscore = rawRuntimeVersion.lastIndexOf('_');
+    String rawRuntimeVersionLastPart;
+    if (lastUnderscore == -1) {
+      rawRuntimeVersionLastPart = rawRuntimeVersion;
+    } else {
+      rawRuntimeVersionLastPart = rawRuntimeVersion.substring(lastUnderscore + 1);
+    }
+    try {
+      runtimeVersion = Integer.parseUnsignedInt(rawRuntimeVersionLastPart);
+    } catch (NumberFormatException ignored) {
+      // Could be e.g. the "jdk" part of "local_jdk" without a version number, which we can't
+      // check at this point.
+      return;
+    }
+
+    if (languageVersion > runtimeVersion) {
+      if (isExec) {
+        throw new InvalidConfigurationException(
+            String.format(
+                "--tool_java_language_version=%s is incompatible with current Java runtime '%s':"
+                    + " Both --tool_java_runtime_version and the 'java_runtime' attribute of the"
+                    + " Java toolchain have to be set to versions at least as high as the language"
+                    + " version.",
+                rawLanguageVersion, rawRuntimeVersion));
+      } else {
+        throw new InvalidConfigurationException(
+            String.format(
+                "--java_language_version=%s is incompatible with --java_runtime_version=%s: The"
+                    + " runtime version has to be at least as high as the language version.",
+                rawLanguageVersion, rawRuntimeVersion));
+      }
+    }
+  }
```

Comment:
If we go with the logic in this class, it would be nice to have some unit test coverage for the parsing logic in `JavaConfigurationTest`